ouroboros-network/src/Data/Sequence/Strict.hs

```diff
@@ -0,0 +1,226 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE ViewPatterns #-}
+
+-- | Strict variants of 'Seq' operations.
+--
+-- TODO: Move this to @cardano-prelude@ as a new module, @Data.Sequence.Strict@.
+module Data.Sequence.Strict
+  ( StrictSeq (Empty, (:<|), (:|>))
+
+    -- * Construction
+  , empty
+  , singleton
+  , (<|)
+  , (|>)
+  , (><)
+  , fromList
+  , toStrict
+
+    -- * Deconstruction
+    -- | Additional functions for deconstructing sequences are available
+    -- via the 'Foldable' instance of 'Seq'.
+
+    -- ** Queries
+  , null
+  , length
+
+    -- * Scans
+  , scanl
+
+    -- * Sublists
+
+    -- ** Sequential searches
+  , dropWhileL
+  , dropWhileR
+  , spanl
+
+    -- * Indexing
+  , take
+  , drop
+  , splitAt
+  ) where
+
+import           Prelude hiding (drop, length, null, scanl, splitAt, take)
+import           Cardano.Prelude (NoUnexpectedThunks (..), forceElemsToWHNF,
+                     noUnexpectedThunksInValues)
+
+import           Codec.Serialise (Serialise)
+import           Data.Foldable (foldl', toList)
+import           Data.Sequence (Seq)
+import qualified Data.Sequence as Seq
+
+infixr 5 ><
+infixr 5 <|
+infixl 5 |>
+
+infixr 5 :<|
+infixl 5 :|>
+
+{-# COMPLETE Empty, (:<|) #-}
+{-# COMPLETE Empty, (:|>) #-}
+
+-- | A @newtype@ wrapper around a 'Seq', representing a general-purpose finite
+-- sequence that is strict in its values.
+--
+-- This strictness is not enforced at the type level, but rather by the
+-- construction functions in this module. These functions essentially just
+-- wrap the original "Data.Sequence" functions while forcing the provided
+-- value to WHNF.
+newtype StrictSeq a = StrictSeq (Seq a)
+  deriving (Eq, Ord, Show)
+  deriving newtype (Foldable, Semigroup, Serialise)
+
+-- | Instance for 'StrictSeq' checks elements only
+--
+-- The internal fingertree in 'Seq' might have thunks, which is essential for
+-- its asymptotic complexity.
+instance NoUnexpectedThunks a => NoUnexpectedThunks (StrictSeq a) where
+  showTypeOf _ = "StrictSeq"
+  whnfNoUnexpectedThunks ctxt = noUnexpectedThunksInValues ctxt . toList
+
+-- | A helper function for the ':<|' pattern.
+--
+viewFront :: StrictSeq a -> Maybe (a, StrictSeq a)
+viewFront (StrictSeq xs) = case Seq.viewl xs of
+                          Seq.EmptyL   -> Nothing
+                          x Seq.:< xs' -> Just (x, StrictSeq xs')
```

Comment:
Style nazi: I'm not sure what you're aligning these patterns with. I would indent them 4 spaces from the start of the line. Alternatively, indent them 4 (or 2) spaces w.r.t. `case`, but that shifts it so much to the right...

 Same for `viewBack`.